hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java

```diff
@@ -809,6 +882,246 @@ public Bytes getFailureResultFor(final ResponseCodeEnum status) {
+			 * check if each one specifies the type of key it applies to (that the bit field
+			 * is not `00000000` and no bit bigger than 6 is set) and also that there are not multiple
+			 * keys values for the same key type (e.g. multiple `TokenKey` instances have the adminKey bit set)
+			 */
+			final var tokenKeys = tokenCreateOp.getTokenKeys();
+			if (!tokenKeys.isEmpty()) {
+				for (int i = 0, tokenKeysSize = tokenKeys.size(); i < tokenKeysSize; i++) {
+					final var tokenKey = tokenKeys.get(i);
+					validateTrue(tokenKey.key().getKeyValueType() != INVALID_KEY, INVALID_TRANSACTION_BODY);
+					final var tokenKeyBitField = tokenKey.keyType();
+					validateTrue(tokenKeyBitField != 0 && tokenKeyBitField < 128, INVALID_TRANSACTION_BODY);
+					for (int j = i + 1; j < tokenKeysSize; j++) {
+						validateTrue((tokenKeyBitField & tokenKeys.get(j).keyType()) == 0,
+								INVALID_TRANSACTION_BODY);
+					}
+				}
+			}
+
+			/*
+			 * The denomination of a fixed fee depends on the values of tokenId, useHbarsForPayment
+			 * useCurrentTokenForPayment. Exactly one of the values of the struct should be set.
+			 */
+			if (!tokenCreateOp.getFixedFees().isEmpty()) {
+				for (final var fixedFee: tokenCreateOp.getFixedFees()) {
+					validateTrue(fixedFee.getFixedFeePayment() != INVALID_PAYMENT, INVALID_TRANSACTION_BODY);
+				}
+			}
+
+			/*
+			 * When a royalty fee with fallback fee is specified, we need to check that
+			 * the fallback fixed fee is valid.
+			 */
+			if (!tokenCreateOp.getRoyaltyFees().isEmpty()) {
+				for (final var royaltyFee: tokenCreateOp.getRoyaltyFees()) {
+					if (royaltyFee.fallbackFixedFee() != null) {
+						validateTrue(royaltyFee.fallbackFixedFee().getFixedFeePayment() != INVALID_PAYMENT,
+								INVALID_TRANSACTION_BODY);
+					}
+				}
+			}
+		}
+
+		private void replaceInheritedKeysWithSenderKey() throws DecoderException {
+			tokenCreateOp.setAllInheritedKeysTo((JKey) ledgers.accounts().get(
+					EntityIdUtils.accountIdFromEvmAddress(senderAddress), AccountProperty.KEY));
+		}
+
+		private boolean validateAdminKey(
+				final MessageFrame frame,
+				final TokenCreateWrapper.TokenKeyWrapper tokenKeyWrapper
+		) {
+			final var key = tokenKeyWrapper.key();
+			return switch (key.getKeyValueType()) {
+				case INHERIT_ACCOUNT_KEY -> validateKey(frame, senderAddress, sigsVerifier::hasActiveKey);
+				case CONTRACT_ID -> validateKey(frame, asTypedEvmAddress(key.getContractID()),
+								sigsVerifier::hasActiveKey);
+				case DELEGATABLE_CONTRACT_ID -> validateKey(frame, asTypedEvmAddress(key.getDelegatableContractID()),
+								sigsVerifier::hasActiveKey);
+				case ED25519 -> validateCryptoKey(new JEd25519Key(key.getEd25519Key()),
+								sigsVerifier::cryptoKeyIsActive);
+				case ECDSA_SECPK256K1 -> validateCryptoKey(new JECDSASecp256k1Key(key.getEcdsaSecp256k1()),
+								sigsVerifier::cryptoKeyIsActive);
+				default -> throw new InvalidTransactionException(NOT_SUPPORTED);
```

Comment:
Normally I take code coverage warnings with a grain of salt, but this one seems important.  We need a test with and INVALID_KEY, even a unit test.

Reply:
The thing is the execution should _never_ reach this default case, since we explicitly check each of the keys for validity in the `verifySolidityInput()`, which is called in `body()`. If any of them is `INVALID_KEY`, a valid synthetic transaction body cannot be constructed from the input, so we revert and never reach this execution point.

I don't really see a way to test this, we do not have a seam, we have to go through the whole flow of the precompile and the execution will fail before we reach this case. (I guess _technically_ we can mock the key to return an `INVALID_KEY` the second time it is queried, but that test would be rather pointless and would serve solely for increasing coverage)?

Maybe we can make the exception reason `FAIL_INVALID`, since reaching it (not really sure how) would mean a bug/internal failure/exceptional circumstance, as Stoyan and Michael noted above?

Reply:
Maybe it makes sense to mock the result from `verifySolidityInput()` just for the purpose of testing `validateAdminKey()`, this way we can make sure that any potential future changes in `verifySolidityInput()` won't cause any breaking issues. 
> Maybe we can make the exception reason FAIL_INVALID, since reaching it (not really sure how) would mean a bug/internal failure/exceptional circumstance, as Stoyan and Michael noted above?

Not sure about `FAIL_INVALID` here, I'm considering just returning `false` from the switch which will then result in line 980 throwing `INVALID_SIGNATURE`.